eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java

```diff
@@ -147,6 +147,30 @@ default T getOnly()
 
         throw new IllegalStateException("Size must be 1 but was " + this.size());
     }
+    
+    /**
+     * Returns the element as an optional if the iterable has exactly one element.
+     * If the iterable is empty {@link Optional#empty} is returned.
+     * Otherwise, throw {@link IllegalStateException}.
+     *
+     * @return an element of an iterable wrapped in an optional or an empty optional.
+     * @throws IllegalStateException if iterable has more than one element.
```

Comment:
I wrote `Iterate.getOnly()` and I've been using it for a while now, so I feel like I can criticize it. The methods which may throw, like `getOnly()` and `groupByUniqueKey()` can be pretty difficult to use, not so much because they throw, but because the type of exception thrown and the error messages they contain are not configurable. Using `Optional` as a return type can help, but in my opinion we should only return `Optional` and never throw.

```java
default Optional<T> getOnlyOptional()
{
    if (this.size() == 1)
    {
        return Optional.ofNullable(this.getFirst());
    }
    return Optional.empty();
}
```

This isn't perfect because it's still hard to throw an error message differentiating between empty and 2-or-more, but I think it's less surprising. An `Either` type would help more. Overall though I think it's surprising for a method which returns `Optional` or `Either` to ever throw.